src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilder.cs

```diff
@@ -38,9 +38,13 @@ public HandlerMiddlewareBuilder(IHandlerResolver handlerResolver, IServiceResolv
     /// </summary>
     /// <typeparam name="TMiddleware">The type of the middleware to add.</typeparam>
     /// <returns>The current HandlerMiddlewareBuilder.</returns>
+    /// <exception cref="InvalidOperationException">When the middleware is not registered as Transient, an exception will be thrown if the resolved middleware is already part of a pipeline.</exception>
     public HandlerMiddlewareBuilder Use<TMiddleware>() where TMiddleware : MiddlewareBase<HandleMessageContext, bool>
     {
-        _middlewares.Add(() => ServiceResolver.ResolveService<TMiddleware>());
+        var newMiddleware = ServiceResolver.ResolveService<TMiddleware>();
+        if (newMiddleware.HasNext) throw new InvalidOperationException("Middlewares must be registered as Transient.");
```

Comment:
Is calling this "Transient" maybe a leaking implementation detail, or is it the right catch-all term?

Reply:
Yeah it's a good question - the point here is that we're indicating to users that they've registered a custom middleware incorrectly. Autofac uses `InstancePerDependency`, while StructureMap/MSDI uses `Transient`, so I thought the latter might be more widely used?

Reply:
Maybe we could be a little bit more verbose in the exception message? I'm not against the way it is, just wondering if there's a different wording we could use to convey the idea and it still fitting other DI containers' concepts/terminology.

Reply:
How about:


Middlewares must be registered into your DI container so that each resolution creates a new instance. For StructureMap use `Transient()`, and for Microsoft.Extensions.DependencyInjection, use `AddTransient`. Please check the documentation for your container for more details.

Reply:
I like it.